src/state-manager/TransactionConsensus.ts

```diff
@@ -1024,6 +1094,565 @@ class TransactionConsenus {
+
+          const executionGroupNodes = new Set(queueEntry.executionGroup.map((node) => node.publicKey))
+          const hasTwoThirdsMajority = this.verifyAppliedReceipt(payload, executionGroupNodes)
+          if (!hasTwoThirdsMajority) {
+            /* prettier-ignore */ if (logFlags.error) this.mainLogger.error(`Receipt does not have the required majority for txid: ${payload.txid}`)
+            nestedCountersInstance.countEvent('poqo', 'poqo-receipt-gossip: Rejecting receipt because no majority')
+            return
+          }
+          
+          queueEntry.poqoReceipt = payload
+          queueEntry.appliedReceipt2 = payload
+          queueEntry.recievedAppliedReceipt2 = payload
+          Comms.sendGossip(
+            'poqo-receipt-gossip',
+            payload,
+            null,
+            null,
+            queueEntry.transactionGroup,
+            false,
+            4,
+            payload.txid,
+            '',
+            true
+          )
+          queueEntry.hasSentFinalReceipt = true
+        } finally {
+          profilerInstance.scopedProfileSectionEnd('poqo-receipt-gossip')
+        }
+      }
+    )
+
+    const poqoDataAndReceiptBinaryHandler: Route<InternalBinaryHandler<Buffer>> = {
+      name: InternalRouteEnum.binary_poqo_data_and_receipt,
+      handler: async (payload, respond, header, sign) => {
+        const route = InternalRouteEnum.binary_poqo_data_and_receipt
+        this.profiler.scopedProfileSectionStart(route, false)
+        try {
+          const _sender = header.sender_id
+          const reqStream = getStreamWithTypeCheck(payload, TypeIdentifierEnum.cPoqoDataAndReceiptReq)
+          if (!reqStream) {
+            nestedCountersInstance.countEvent('internal', `${route}-invalid_request`)
+            return
+          }
+          const readableReq = deserializePoqoDataAndReceiptResp(reqStream)
+          // make sure we have it
+          const queueEntry = this.stateManager.transactionQueue.getQueueEntrySafe(readableReq.finalState.txid) // , payload.timestamp)
+          //It is okay to ignore this transaction if the txId is not found in the queue.
+          if (queueEntry == null) {
+            //In the past we would enqueue the TX, expecially if syncing but that has been removed.
+            //The normal mechanism of sharing TXs is good enough.
+            nestedCountersInstance.countEvent('processing', 'broadcast_finalstate_noQueueEntry')
+            return
+          }
+
+          // validate corresponding tell sender
+          if (_sender == null) {
+            /* prettier-ignore */ if (logFlags.error) this.mainLogger.error(`poqo-data-and-receipt invalid sender for txid: ${readableReq.finalState.txid}, sender: ${_sender}`)
+            return
+          }
+
+          const isValidFinalDataSender =
+            this.stateManager.transactionQueue.factValidateCorrespondingTellFinalDataSender(
+              queueEntry,
+              _sender
```

Comment:
not sure that it was fixed, at least I didn't find - _sender in header and actual sender who signed message can be different

Reply:
this looks to be covered, but I will get confirmation.
I also have another concern that I thought we had to allow data from both unrwrapped and wrapped index values.   if we don't allow that nodes could be blocking final data that is valid

Reply:
going to leave unresolved but not a merge blocker

Reply:
@arhamj will put in a fix for this today.

Reply:
https://github.com/shardeum/shardus-core/pull/226